### ISSUE TITLE
How to add query data for LightGBM with objective lambdarank

### ISSUE THREAD

**Issue by ttpro1995** [Azure/mmlspark#732]
I am trying to reproduce experience in [Yahoo LTR](https://github.com/guolinke/boosting_tree_benchmarks).

I preprocess with the script [yahoo2libsvm.py](https://github.com/guolinke/boosting_tree_benchmarks/blob/master/data/yahoo2libsvm.py)

It create 6 file, 3 file per train/test 

- yahoo.train
- yahoo.train.group
- yahoo.train.query
- yahoo.test
- yahoo.test.group
- yahoo.test.query

As mention in [LightGBM docs  ](https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst#query-data), it should automatically looking for  `yahoo.train.query` if training data is `yahoo.train`. 

However, what about Spark code in mmlspark ? How do I make use of .query and .group data ? 

**My Spark code** , it is clearly missed something. I am newbie, I don't know. 

```
val parent_folder = "/data/public/dataset/Learning_to_Rank_Challenge/"
val yahoo_train_path = parent_folder + "yahoo.train"
val yahoo_test_path = parent_folder + "yahoo.test"
val train_df = spark.read.format("libsvm").option("numFeatures", "700").load(yahoo_train_path)

import com.microsoft.ml.spark.lightgbm.LightGBMRegressor
val regressor = new LightGBMRegressor().setLabelCol("label").setFeaturesCol("features").setObjective("lambdarank").setNumLeaves(255).setMinSumHessianInLeaf(100).setNumIterations(1)

val model = regressor.fit(train_df)
```

**Exception**
```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 14.0 failed 1 times, most recent failure: Lost task 0.0 in stage 14.0 (TID 26, localhost, executor driver): java.lang.Exception: Booster call failed in LightGBM with error: Lambdarank tasks require query information
	at com.microsoft.ml.spark.lightgbm.LightGBMUtils$.validate(LightGBMUtils.scala:31)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.createBooster(TrainUtils.scala:135)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.translate(TrainUtils.scala:267)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.trainLightGBM(TrainUtils.scala:411)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
  at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
  at scala.Option.foreach(Option.scala:257)
  at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
  at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
  at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2158)
  at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1035)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
  at org.apache.spark.rdd.RDD.reduce(RDD.scala:1017)
  at org.apache.spark.sql.Dataset$$anonfun$reduce$1.apply(Dataset.scala:1637)
  at org.apache.spark.sql.Dataset$$anonfun$withNewRDDExecutionId$1.apply(Dataset.scala:3349)
  at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
  at org.apache.spark.sql.Dataset.withNewRDDExecutionId(Dataset.scala:3345)
  at org.apache.spark.sql.Dataset.reduce(Dataset.scala:1636)
  at com.microsoft.ml.spark.lightgbm.LightGBMBase$class.innerTrain(LightGBMBase.scala:90)
  at com.microsoft.ml.spark.lightgbm.LightGBMRegressor.innerTrain(LightGBMRegressor.scala:38)
  at com.microsoft.ml.spark.lightgbm.LightGBMBase$class.train(LightGBMBase.scala:38)
  at com.microsoft.ml.spark.lightgbm.LightGBMRegressor.train(LightGBMRegressor.scala:38)
  at com.microsoft.ml.spark.lightgbm.LightGBMRegressor.train(LightGBMRegressor.scala:38)
  at org.apache.spark.ml.Predictor.fit(Predictor.scala:118)
  ... 47 elided
Caused by: java.lang.Exception: Booster call failed in LightGBM with error: Lambdarank tasks require query information
  at com.microsoft.ml.spark.lightgbm.LightGBMUtils$.validate(LightGBMUtils.scala:31)
  at com.microsoft.ml.spark.lightgbm.TrainUtils$.createBooster(TrainUtils.scala:135)
  at com.microsoft.ml.spark.lightgbm.TrainUtils$.translate(TrainUtils.scala:267)
  at com.microsoft.ml.spark.lightgbm.TrainUtils$.trainLightGBM(TrainUtils.scala:411)
  at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
  at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
  at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
  at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
  at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
  at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
  at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
  at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
  at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
  at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
  at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
  at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
  at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
  at org.apache.spark.scheduler.Task.run(Task.scala:121)
  at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
  at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
  ... 3 more
```

**Comment by imatiach-msft** [comment]
@ttpro1995 there's just a query column in lightgbm, where you specify for each row what group it is in.  The lightgbm query data is more compressed in that it specifies the count of items in that group.  I wasn't sure how to add that kind of format to a spark dataframe, since the row count would be different from the dataset (there's no 'jagged' columns in spark).  Moreover, relying on row order is very dangerous in spark since whenever a shuffle occurs the ordering of rows changes, eg see:
https://stackoverflow.com/questions/45822242/does-spark-preserve-record-order-when-reading-in-ordered-files
however I do a within-partition sort prior to training

**Comment by ttpro1995** [comment]
Please add code example, prefer Spark Scala.

**Comment by imatiach-msft** [comment]
@ttpro1995 
I think this might be a good scala example:
https://github.com/Azure/mmlspark/blob/master/src/test/scala/com/microsoft/ml/spark/lightgbm/split2/VerifyLightGBMRanker.scala#L39

**Comment by ranadeepsingh** [comment]
Thank you for reaching out! To perform ranking tasks like lambdarank in SynapseML, please use `LightGBMRanker` instead of `LightGBMRegressor` and specify the group/query column using `.setGroupCol(...)`. The maintainers provided sample code demonstrating this usage in the discussion thread. We are closing this issue as the question has been answered, but feel free to open a new issue if you have further questions.
